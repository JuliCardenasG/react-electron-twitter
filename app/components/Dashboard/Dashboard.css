.container {
  background-color: #ffffff;
  width: 70%;
}
