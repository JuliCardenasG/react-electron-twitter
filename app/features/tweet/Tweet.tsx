import React, { ReactElement } from 'react';
import { Card, Avatar } from 'antd';
import { TweetType } from './tweet';


export default function Tweet(tweet: TweetType): ReactElement {
  const { Meta } = Card;

  return (
    <Card>
      <Meta
        avatar={<Avatar src={tweet.user.profile_image_url_https} />}
        title={tweet.user.screen_name}
        description={tweet.text}
      />
    </Card>
  )
}
