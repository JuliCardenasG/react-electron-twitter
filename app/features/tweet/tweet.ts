// File for Twitter related types
export type TweetType = {
  created_at: string;
  id: number;
  id_str: string;
  text: string;
  retweet_count: number;
  favorite_count: number;
  user: UserType;
};

export type UserType = {
  id: number;
  id_str: string;
  name: string;
  screen_name: string;
  location: string;
  description: string;
  url: string;
  protected: boolean;
  friends_count: number;
  followers_count: number;
  verified: boolean;
  profile_image_url_https: string;
  profile_banner_url: string;
}
