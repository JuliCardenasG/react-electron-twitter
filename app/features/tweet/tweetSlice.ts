import { createSlice } from '@reduxjs/toolkit';
import { TweetType, UserType } from './tweet';

export const tweetSlice = createSlice({
  name: 'tweet',
  initialState: {
    username: '',
    currentUser: {} as UserType,
    tweets: [] as TweetType[],
    isLoading: false,
    error: false
  },
  reducers: {
    tweetsLoad: (state) => {
      state.tweets = [],
      state.isLoading = true;
      state.error = false
    },
    tweetsSuccess: (state, action) => {
      state.tweets = action.payload;
      state.isLoading = false;
      state.error = false;
    },
    tweetsError: (state, action) => {
      state.isLoading = false;
      state.error = action.payload;
    },
    setUsername: (state, action) => {
      state.username = action.payload;
      state.isLoading = false;
    },
    setCurrentUser: (state, action) => {
      state.currentUser = action.payload;
      state.isLoading = false;
    },
  }
})

export default tweetSlice.reducer;

