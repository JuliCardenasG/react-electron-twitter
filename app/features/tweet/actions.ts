// Actions for Tweet slice

import { api } from '../../api';
import { tweetSlice } from './tweetSlice';

const { tweetsLoad, tweetsSuccess, tweetsError, setUsername } = tweetSlice.actions

// Fetch tweets from Twitter API
export const fetchTweets = (twitterUsername: string) => async dispatch => {
  try {
    dispatch(tweetsLoad());

    // Params for searching tweets for given username
    const params = {
      screen_name: twitterUsername
    }

    await api.get('/statuses/user_timeline.json', { params })
      .then((response) => dispatch(tweetsSuccess(response.data)))
  }
  catch (e) {
    dispatch(tweetsError(e.message));
    return console.error(e.message);
  }
}

// Set Twitter username for timeline search
export const setUsernameAction = (twitterUsername: string) => async dispatch => {
  dispatch(setUsername(twitterUsername));
}
