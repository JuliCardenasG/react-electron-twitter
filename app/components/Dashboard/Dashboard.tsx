import { Spin } from 'antd';
import React, { ReactElement, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchTweets } from '../../features/tweet/actions';
import { RootState } from '../../store';
import Header from '../Header/Header';
import Timeline from '../Timeline/Timeline';
import UserInfo from '../UserInfo/UserInfo';
import styles from './Dashboard.css';



export default function Dashboard(): ReactElement {

  const dispatch = useDispatch();
  const { username, tweets, isLoading , currentUser } = useSelector((state: RootState) => state.tweets);

  useEffect(() => {
    dispatch(fetchTweets(username));
  }, [dispatch]);

  const timeLineProps = { tweets, isLoading };
  const userInfoProps = { currentUser };

  return (
    isLoading ?
      <Spin size="large" /> :
      <div className={styles.container}>
        <Header />
        <UserInfo {...userInfoProps}/>
        <Timeline {...timeLineProps}/>
      </div>
  )
}
