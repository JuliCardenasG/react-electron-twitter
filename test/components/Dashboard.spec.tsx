import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { TweetType, UserType } from '../../app/features/tweet/tweet';
import { configureStore } from '@reduxjs/toolkit';
import * as tweetSlice from '../../app/features/tweet/tweetSlice';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import Dashboard from '../../app/components/Dashboard/Dashboard';
import renderer from 'react-test-renderer';
import * as redux from "react-redux";

Enzyme.configure({ adapter: new Adapter() });
jest.useFakeTimers();

//Mocking window.matchMedia as it's not available in the test environment
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

const initialState = {
  username: '',
  currentUser: {} as UserType,
  tweets: [] as TweetType[],
  isLoading: false,
  error: false
}

// Mock data for user
const mockUser: UserType = {
  id: 68653811,
  id_str: "68653811",
  name: "John Doe",
  screen_name: "JohnDoe",
  location: "",
  description: "Description.",
  url: null,
  protected: false,
  followers_count: 195,
  friends_count: 155,
  verified: false,
  profile_image_url_https: "https://pbs.twimg.com/profile_images/766422613182783488/CIPmI1mj_normal.jpg",
  profile_banner_url: "https://pbs.twimg.com/profile_banners/68653811/1469288403",
}

// Mock tweets to test timeline rendering
const mockTweets: TweetType[] = [
  {
    created_at: "Sat Nov 07 19:54:24 +0000 2020",
    id: 132516472105881190,
    id_str: "132516472105881190",
    text: "Lorem Ipsum Dolor Sit Amet",
    retweet_count: 15,
    favorite_count: 10,
    user: {...mockUser},
  },
  {
    created_at: "Fri Nov 06 14:05:08 +0000 2020",
    id: 1324714436897230850,
    id_str: "1324714436897230850",
    text: "Amet Lorem Ipsum Dolor Sit ",
    retweet_count: 5,
    favorite_count: 2,
    user: {...mockUser},
  },
];

const tweetsPreloadState =  {
  tweets: { ...initialState}
}

function setup (preloadedState = tweetsPreloadState) {
  const store = configureStore({
    reducer: { tweets: tweetSlice.default },
    preloadedState
  });
  const getWrapper = () => {
    mount(
      <Provider store={store}>
        <Router>
          <Dashboard />
        </Router>
      </Provider>
    )
  }
  const shallowComponent = shallow(
    <Provider store={store}>
        <Router>
          <Dashboard />
        </Router>
      </Provider>
  )
  const component = getWrapper();
  return { store, component, shallowComponent };
}

describe('Dashboard component', () => {
  it('should render', () => {
    const { shallowComponent } = setup({
      tweets: {
        username: mockUser.screen_name,
        currentUser: { ...mockUser },
        tweets: [...mockTweets],
        ...initialState
      }
    }, );
    expect(shallowComponent).toBeTruthy();
  });

  it('should match exact snapshot', () => {
    const { store } = setup();
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Dashboard />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should dispatch fetch tweets actions', () => {
    // Mocke useDispatch function
    const useDispatchSpy = jest.spyOn(redux, 'useDispatch');
    const mockDispatchFn = jest.fn();
    useDispatchSpy.mockReturnValue(mockDispatchFn);

    const { shallowComponent } = setup({
      tweets: {
        username: mockUser.screen_name,
        currentUser: { ...mockUser },
        tweets: [...mockTweets],
        ...initialState
      }
    }, );
    expect(mockDispatchFn).toHaveBeenCalled();

    useDispatchSpy.mockClear();
  });
})
