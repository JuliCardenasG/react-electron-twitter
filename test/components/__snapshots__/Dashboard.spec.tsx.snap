// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component should match exact snapshot 1`] = `
<div
  className="ant-spin ant-spin-lg ant-spin-spinning"
>
  <span
    className="ant-spin-dot ant-spin-dot-spin"
  >
    <i
      className="ant-spin-dot-item"
    />
    <i
      className="ant-spin-dot-item"
    />
    <i
      className="ant-spin-dot-item"
    />
    <i
      className="ant-spin-dot-item"
    />
  </span>
</div>
`;
